src/staking/observable/index.ts

```diff
@@ -0,0 +1,372 @@
+}
+
+/**
+ * ObservableStaking is a class that provides an interface to create observable
+ * staking transactions for the Babylon Staking protocol.
+ * 
+ * The class requires a network and staker information to create staking
+ * transactions.
+ * The staker information includes the staker's address and 
+ * public key(without coordinates).
+ */
+export class ObservableStaking {
+  private network: networks.Network;
+  private stakerInfo: StakerInfo;
+
+  constructor(network: networks.Network, stakerInfo: StakerInfo) {
+    if (!isValidBitcoinAddress(stakerInfo.address, network)) {
+      throw new StakingError(
+        StakingErrorCode.INVALID_INPUT, "Invalid staker bitcoin address",
+      );
+    }
+    if (!isValidNoCordPublicKey(stakerInfo.publicKeyNoCoordHex)) {
+      throw new StakingError(
+        StakingErrorCode.INVALID_INPUT, "Invalid staker public key",
+      );
+    }
+    this.network = network;
+    this.stakerInfo = stakerInfo;
+  }
+
+  /**
+   * Create a staking transaction for observable staking.
+   * 
+   * @param {ObservableStakingParams} params - The staking parameters for observable staking.
+   * @param {number} stakingAmountSat - The amount to stake in satoshis.
+   * @param {number} stakingTimeBlocks - The time to stake in blocks.
+   * @param {string} finalityProviderPkNoCoord - The finality provider's public key
+   * without coordinates.
+   * @param {UTXO[]} inputUTXOs - The UTXOs to use as inputs for the staking 
+   * transaction.
+   * @param {number} feeRate - The fee rate for the transaction in satoshis per byte.
+   * 
+   * @returns {PsbtTransactionResult} - An object containing the unsigned psbt and fee
+   */
+  public createStakingTransaction = (
+    params: ObservableStakingParams,
+    stakingAmountSat: number,
+    stakingTimeBlocks: number,
+    finalityProviderPkNoCoord: string,
+    inputUTXOs: UTXO[],
+    feeRate: number,
+  ): PsbtTransactionResult => {
+    validateParams(params);
+    const stakingTerm = getStakingTerm(params, stakingTimeBlocks);
+
+    validateStakingTxInputData(
+      stakingAmountSat,
+      stakingTerm,
+      params,
+      inputUTXOs,
+      feeRate,
+    );
+
+    const scripts = buildScripts(
+      params,
+      finalityProviderPkNoCoord,
+      stakingTerm,
+      this.stakerInfo.publicKeyNoCoordHex,
+    );
+
+    // Create the staking transaction
+    try {
+      return stakingTransaction(
+        scripts,
+        stakingAmountSat,
+        this.stakerInfo.address,
+        inputUTXOs,
+        this.network,
+        feeRate,
+        isTaproot(this.stakerInfo.address, this.network) ? Buffer.from(this.stakerInfo.publicKeyNoCoordHex, "hex") : undefined,
+        // `lockHeight` is exclusive of the provided value.
+        // For example, if a Bitcoin height of X is provided,
+        // the transaction will be included starting from height X+1.
+        // https://learnmeabitcoin.com/technical/transaction/locktime/
+        params.activationHeight - 1,
```

Comment:
hmm is there an option for the caller to not provide `locktime` ?

Reply:
hmm I am a bit on the fence here. 

From protocol point of view we do not enforce `timelock` field to have any concrete value i.e staking transaction can have arbitrary timelock and be valid staking transaction. So I imagine somebody could want to construct staking transaction with `timelock` equal to `0` i.e no timelock.

So this field being equal to `activationHeight` is our implementation detail which we are forcing on our callers.

That said, as I understand this field is important from the point of view of the consumers of the library, so that they will always construct staking transaction which corresponds to correct params. Is that right ?